### PR TITLE
Add sauceball trackball

### PR DESCRIPTION
**Pull request by xia0**
Add support for sauceball trackball

## Description

a 3d printed trackball which aims to convert ps/2 optical mice into trackballs.
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/sauceball/readme.md
Comment (keyboard-magpie) [inline keyboards/sauceball/readme.md:27]: Can you update this to actually reflect your hardware please, is there actually a reset button?

**Review by keyboard-magpie — CHANGES_REQUESTED** [review]

File: keyboards/sauceball/keymaps/default/keymap.c
Comment (keyboard-magpie) [inline keyboards/sauceball/keymaps/default/keymap.c:1]: please include a license header here.

**Review by zvecr — CHANGES_REQUESTED** [review]
Should be moved to the `keyboards/handwired/ folder.

File: keyboards/sauceball/readme.md
Comment (xia0) [reply keyboards/sauceball/readme.md:27]: The default message already seems pretty accurate. The Teensy 2.0 has a reset button but the Arduino Pro Micros have pads that need to be shorted.

File: keyboards/handwired/sauceball/info.json
Comment (keyboard-magpie) [inline keyboards/handwired/sauceball/info.json:5]: If your hardware section is talking about teensy vs pro micro, this is then incorrect.

File: keyboards/handwired/sauceball/readme.md
Comment (xia0) [reply keyboards/handwired/sauceball/readme.md:27]: Understood. Updated readme file accordingly.

File: keyboards/handwired/sauceball/info.json
Comment (drashna) [reply keyboards/handwired/sauceball/info.json:5]:
```suggestion
    "bootloader": "halfkay",
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/handwired/sauceball/keymaps/default/keymap.c:19]: The default keymap should be kept simple and minimal. 

This should be moved into another keymap, or into the keyboard's c file (eg, sauceball.c)

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/default/keymap.c:19]: Will do. I'm away for a few weeks so I'll do it when I get back and can test on hardware. I'll revert this PR to draft.

Comment (drashna) [reply keyboards/handwired/sauceball/keymaps/default/keymap.c:19]: I'll add an "on hold" label for you. Let us know when you have it working.

**Review by drashna — APPROVED** [review]

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/default/keymap.c]: > I'll add an "on hold" label for you. Let us know when you have it working.

The default keymap works fine on hardware.

The xia0 keymap works too but I've disabled a custom key for drag-scrolling for the time being because I cannot figure out why tap_code(KC_BTN#) does not work but works for other keycodes within process_record_user.

**Comment by xia0** [comment]
@drashna tested and working. removal of on hold label much appreciated. thank you

File: keyboards/handwired/sauceball/keymaps/xia0/keymap.c
Comment (drashna) [inline keyboards/handwired/sauceball/keymaps/xia0/keymap.c:9]:
```suggestion
  NO_SCROLL = QK_USER,
```

File: keyboards/handwired/sauceball/keymaps/xia0/keymap.c
Comment (fauxpark) [inline keyboards/handwired/sauceball/keymaps/xia0/keymap.c:6]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/sauceball/keymaps/xia0/keymap.c
Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/xia0/keymap.c:6]: @fauxpark on line 57 and onwards the functions mousekey_clear, mousekey_on and mousekey_off are used. Would these be affected if mousekey.h is not included?

Comment (fauxpark) [reply keyboards/handwired/sauceball/keymaps/xia0/keymap.c:6]: It should be implicitly included via `QMK_KEYBOARD_H` (default_keyboard.h) <- quantum.h
If not, that's a bug.

Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/xia0/keymap.c:6]: @fauxpark didn't build after removing the include with the following error:

> Compiling: quantum/keymap_introspection.c                                                          In file included from quantum/keymap_introspection.c:5:
> ./keyboards/handwired/sauceball/keymaps/xia0/keymap.c: In function 'process_record_user':
> ./keyboards/handwired/sauceball/keymaps/xia0/keymap.c:54:11: error: implicit declaration of function 'mousekey_clear'; did you mean 'timer_clear'? [-Werror=implicit-function-declaration]
>            mousekey_clear();
>            ^~~~~~~~~~~~~~
>            timer_clear
> ./keyboards/handwired/sauceball/keymaps/xia0/keymap.c:55:11: error: implicit declaration of function 'mousekey_on' [-Werror=implicit-function-declaration]
>            mousekey_on(DRAG_SCROLL_KEY);
>            ^~~~~~~~~~~
> ./keyboards/handwired/sauceball/keymaps/xia0/keymap.c:56:11: error: implicit declaration of function 'mousekey_send'; did you mean 'host_system_send'? [-Werror=implicit-function-declaration]
>            mousekey_send();
>            ^~~~~~~~~~~~~
>            host_system_send
> ./keyboards/handwired/sauceball/keymaps/xia0/keymap.c:58:11: error: implicit declaration of function 'mousekey_off' [-Werror=implicit-function-declaration]
>            mousekey_off(DRAG_SCROLL_KEY);
>            ^~~~~~~~~~~~
> cc1: all warnings being treated as errors
>  [ERRORS]
>  |
>  |
>  |
> gmake[1]: *** [builddefs/common_rules.mk:357: .build/obj_handwired_sauceball_xia0/quantum/keymap_introspection.o] Error 1
> gmake[1]: *** Waiting for unfinished jobs....

Just gonna revert the changes for now.

Comment (fauxpark) [reply keyboards/handwired/sauceball/keymaps/xia0/keymap.c:6]: Yeah, from a quick sweep of the repo it seems `mousekey.h` is not included anywhere it would get picked up through quantum.h. I'll try to fix that up in a bit.

**Comment by xia0** [comment]
Hello, any chance I can get this an awaiting review label? Cheers

File: keyboards/handwired/sauceball/config.h
Comment (waffle87) [inline keyboards/handwired/sauceball/config.h:21]:
```suggestion
```

File: keyboards/handwired/sauceball/info.json
Comment (waffle87) [inline keyboards/handwired/sauceball/info.json:29]:
```suggestion
        "LAYOUT": {
```

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (waffle87) [inline keyboards/handwired/sauceball/keymaps/default/keymap.c:7]:
```suggestion
  [0] = LAYOUT(KC_BTN1, KC_BTN3, KC_BTN4, KC_BTN2)
```

File: keyboards/handwired/sauceball/keymaps/xia0/keymap.c
Comment (waffle87) [inline keyboards/handwired/sauceball/keymaps/xia0/keymap.c:13]:
```suggestion
  [0] = LAYOUT(KC_BTN1, KC_BTN3, NO_SCROLL, KC_BTN2)
```

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/handwired/sauceball/keymaps/default/keymap.c:1]: in general, we want the default keymap to be pristine, and simple. 

You can, however, move a lot of this a "manufacturer" keymap with the more complicated code there. 


PR Checklest:
https://docs.qmk.fm/#/pr_checklist?id=keyboard-prs

> - default (and via) keymaps should be “pristine”
>   - bare minimum to be used as a “clean slate” for another user to develop their own user-specific keymap
>   - what does pristine mean? no custom keycodes. no advanced features like tap dance or macros. basic mod taps and home row mods would be acceptable where their use is necessary
>   - standard layouts preferred in these keymaps, if possible
>   - should use [encoder map feature](https://docs.qmk.fm/#/feature_encoders?id=encoder-map), rather than encoder_update_user()
> - submitters can have a "bells-and-whistle") keymap showcasing capabilities in the same PR but it shouldn’t be embedded in the ‘default’ keymap
> - submitters can also have a “manufacturer-matching” keymap that mirrors existing functionality of the commercial product, if porting an existing board

Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/default/keymap.c:1]: >         submitters can also have a “manufacturer-matching” keymap that mirrors existing functionality of the commercial product, if porting an existing board

Do you mean a keymap with the name matching the manufacturer in info.json?

File: keyboards/handwired/sauceball/keymaps/xia0/keymap.c
Comment (fauxpark) [reply keyboards/handwired/sauceball/keymaps/xia0/keymap.c]: Oops, I kinda forgot about this. I've opened https://github.com/qmk/qmk_firmware/pull/21897 to fix it.

File: keyboards/handwired/sauceball/keymaps/default/keymap.c
Comment (drashna) [reply keyboards/handwired/sauceball/keymaps/default/keymap.c:1]: The name doesn't have to match the manufacturing name.  It could be "stock", "feature" etc.  Something that is clear and to the point is ideal. 

The reason for this, is that new keymaps are created off of the default keymap, and historically, the more complex the default is, the harder it is for newer users, as it can/does create confusion.  So, we want it to be plain and simple.

File: keyboards/handwired/sauceball/keymaps/default/config.h
Comment (drashna) [inline keyboards/handwired/sauceball/keymaps/default/config.h:4]:
```suggestion
#pragma once

// Sensitivity fine-tuning
```

**Review by drashna — APPROVED** [review]

File: keyboards/handwired/sauceball/keymaps/stock/keymap.c
Comment (fauxpark) [inline keyboards/handwired/sauceball/keymaps/stock/keymap.c:6]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/sauceball/info.json
Comment (zvecr) [inline keyboards/handwired/sauceball/info.json:15]: With the process here being `feature.to_uppercase` + `_ENABLE`, the resulting `DEBUG_MOUSE_ENABLE` and `DEBUG_ENABLE_ENABLE` are not valid options.

```suggestion
        "nkro": true
```

File: keyboards/handwired/sauceball/keymaps/stock/keymap.c
Comment (xia0) [reply keyboards/handwired/sauceball/keymaps/stock/keymap.c]: thank you :') #21897 is merged into develop branch. Does that mean I'll need to wait until its merged into master before this PR will be?

**Review by drashna — APPROVED** [review]

**Pull request closed without merging**
